src/lib/consensus/proof_of_stake.ml

```diff
@@ -413,17 +438,29 @@ module Data = struct
           !t.curr_epoch_snapshot <- v
 
     let seen_slot (t : t) epoch slot =
-      match
-        Tuple2.compare ~cmp1:Epoch.compare ~cmp2:Epoch.Slot.compare
-          !t.last_checked_slot_and_epoch (epoch, slot)
-      with
-      | i when i >= 0 ->
-          `Seen
-      | i when i < 0 ->
-          !t.last_checked_slot_and_epoch <- (epoch, slot) ;
-          `Unseen
-      | _ ->
-          failwith "forall x : int. x >= 0 || x < 0 is impossibly false"
+      let module Table = Public_key.Compressed.Table in
+      let unseens =
+        Table.to_alist !t.last_checked_slot_and_epoch
+        |> List.filter_map ~f:(fun (pk, old_epoch_and_slot) ->
+               match
+                 Tuple2.compare ~cmp1:Epoch.compare ~cmp2:Epoch.Slot.compare
+                   old_epoch_and_slot (epoch, slot)
+               with
+               | i when i >= 0 ->
+                   None
+               | i when i < 0 ->
+                   Table.set !t.last_checked_slot_and_epoch ~key:pk
+                     ~data:(epoch, slot) ;
+                   Some pk
+               | _ ->
+                   failwith
+                     "forall x : int. x >= 0 || x < 0 is impossibly false" )
+      in
+      match unseens with
+      | [] ->
+          `All_seen
+      | nel ->
```

Comment:
What is `nel`?

Reply:
non-empty-list. I hate acronyms, but there is prior art: https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/NonEmptyList.scala

Reply:
I would prefer a better name, but the code is clear enough here that it doesn't matter I guess.